src/item_pocket.cpp

```diff
@@ -0,0 +1,580 @@
+    } );
+    return iter != contents.end() ? &*iter : nullptr;
+}
+
+void item_pocket::casings_handle( const std::function<bool( item & )> &func )
+{
+    for( auto it = contents.begin(); it != contents.end(); ) {
+        if( it->has_flag( "CASING" ) ) {
+            it->unset_flag( "CASING" );
+            if( func( *it ) ) {
+                it = contents.erase( it );
+                continue;
+            }
+            // didn't handle the casing so reset the flag ready for next call
+            it->set_flag( "CASING" );
+        }
+        ++it;
+    }
+}
+
+bool item_pocket::use_amount( const itype_id &it, int &quantity, std::list<item> &used )
+{
+    bool used_item = false;
+    for( auto a = contents.begin(); a != contents.end() && quantity > 0; ) {
+        if( a->use_amount( it, quantity, used ) ) {
+            used_item = true;
+            a = contents.erase( a );
+        } else {
+            ++a;
+        }
+    }
+    return used_item;
+}
+
+bool item_pocket::will_explode_in_a_fire() const
+{
+    if( data->fire_protection ) {
+        return false;
+    }
+    return std::any_of( contents.begin(), contents.end(), []( const item & it ) {
+        return it.will_explode_in_fire();
+    } );
+}
+
+bool item_pocket::detonate( const tripoint &pos, std::vector<item> &drops )
+{
+    const auto new_end = std::remove_if( contents.begin(), contents.end(), [&pos, &drops]( item & it ) {
+        return it.detonate( pos, drops );
+    } );
+    if( new_end != contents.end() ) {
+        contents.erase( new_end, contents.end() );
+        // If any of the contents explodes, so does the container
+        return true;
+    }
+    return false;
+}
+
+bool item_pocket::process( const itype &type, player *carrier, const tripoint &pos, bool activate,
+                           float insulation, const temperature_flag flag )
+{
+    const bool preserves = type.container && type.container->preserves;
+    bool processed = false;
+    for( auto it = contents.begin(); it != contents.end(); ) {
+        if( preserves ) {
+            // Simulate that the item has already "rotten" up to last_rot_check, but as item::rot
+            // is not changed, the item is still fresh.
+            it->set_last_rot_check( calendar::turn );
+        }
+        if( it->process( carrier, pos, activate, type.insulation_factor * insulation, flag ) ) {
+            it = contents.erase( it );
+            processed = true;
+        } else {
+            ++it;
+        }
+    }
+    return processed;
+}
+
+bool item_pocket::legacy_unload( player *guy, bool &changed )
+{
+    contents.erase( std::remove_if( contents.begin(), contents.end(),
+    [guy, &changed]( item & e ) {
+        int old_charges = e.charges;
+        const bool consumed = guy->add_or_drop_with_msg( e, true );
+        changed = changed || consumed || e.charges != old_charges;
+        if( consumed ) {
+            guy->mod_moves( -guy->item_handling_cost( e ) );
```

Comment:
You never check `guy` for being a non-null pointer, so you assume the caller will always pass a valid pointer. Please change the parameter to a reference, so the caller **can** only call this with a valid value and one can not accidentally pass a null pointer.